libs/execution_context/ebpf_maps.c

```diff
@@ -525,100 +555,122 @@ _delete_per_cpu_hash_map(_In_ ebpf_core_map_t* map)
+    if (cpu_id < per_cpu_data->count) {
+        return tables + cpu_id;
     } else {
         return NULL;
     }
 }
 
 static uint8_t*
-_find_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key)
+_find_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, uint32_t cpu_id, _In_ const uint8_t* key)
 {
     uint8_t* value = NULL;
-    ebpf_hash_table_t* table;
+    ebpf_lock_state_t lock_state;
+    ebpf_core_per_cpu_hash_table_t* table;
     if (!map || !key)
         return NULL;
 
-    table = _get_hash_table_for_cpu(map);
+    table = _get_hash_table_for_cpu(map, cpu_id);
     if (!table)
         return NULL;
 
-    if (ebpf_hash_table_find(table, key, &value) != EBPF_SUCCESS) {
+    lock_state = ebpf_lock_lock(&table->lock);
+    if (ebpf_hash_table_find(table->table, key, &value) != EBPF_SUCCESS) {
         value = NULL;
     }
-
+    ebpf_lock_unlock(&table->lock, lock_state);
     return value;
 }
 
 static ebpf_result_t
-_update_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key, _In_ const uint8_t* data)
+_update_per_cpu_hash_map_entry(
+    _In_ ebpf_core_map_t* map, uint32_t cpu_id, _In_ const uint8_t* key, _In_ const uint8_t* data)
 {
     ebpf_result_t result;
-    ebpf_hash_table_t* table;
+    ebpf_core_per_cpu_hash_table_t* table;
     size_t entry_count = 0;
     uint8_t* value;
-    if (!map || !key || !data)
+    ebpf_lock_state_t lock_state;
+
+    if (!map || !key)
         return EBPF_INVALID_ARGUMENT;
 
-    table = _get_hash_table_for_cpu(map);
+    table = _get_hash_table_for_cpu(map, cpu_id);
     if (!table)
         return EBPF_INVALID_ARGUMENT;
 
-    entry_count = ebpf_hash_table_key_count(table);
+    lock_state = ebpf_lock_lock(&table->lock);
+    entry_count = ebpf_hash_table_key_count(table->table);
 
     if ((entry_count == map->ebpf_map_definition.max_entries) &&
-        (ebpf_hash_table_find(table, key, &value) != EBPF_SUCCESS))
+        (ebpf_hash_table_find(table->table, key, &value) != EBPF_SUCCESS))
         result = EBPF_INVALID_ARGUMENT;
     else
-        result = ebpf_hash_table_update(table, key, data);
+        result = ebpf_hash_table_update(table->table, key, data);
+
+    ebpf_lock_unlock(&table->lock, lock_state);
 
     return result;
 }
 
 static ebpf_result_t
-_delete_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key)
+_delete_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, uint32_t cpu_id, _In_ const uint8_t* key)
 {
     ebpf_result_t result;
-    ebpf_hash_table_t* table;
+    ebpf_core_per_cpu_hash_table_t* table;
+    ebpf_lock_state_t lock_state;
+
     if (!map || !key)
         return EBPF_INVALID_ARGUMENT;
 
-    table = _get_hash_table_for_cpu(map);
+    table = _get_hash_table_for_cpu(map, cpu_id);
     if (!table)
         return EBPF_INVALID_ARGUMENT;
 
-    result = ebpf_hash_table_delete(_get_hash_table_for_cpu(map), key);
+    lock_state = ebpf_lock_lock(&table->lock);
```

Comment:
Same inefficiency here, of taking a lock and releasing it twice, where it ought to be lock-free, at least when cpu_id is the current CPU.  You can use SAL annotation to note it's at dispatch IRQL, and use current CPU with no lock uses, to keep hot path fast.   If you need to touch the map for a different CPU you can queue a DPC to that CPU since that should not be needed in any hot path.

Reply:
Need to sync with bpf_update_elem which can insert elements from another CPU, hence the need for the lock.